src/validate.rs

```diff
@@ -1,23 +1,723 @@
+    /// These flags are stored in the DNSKEY record.
+    flags: u16,
+
+    /// The raw public key.
+    ///
+    /// This identifies the key and can be used for signatures.
+    key: RawPublicKey,
+}
+
+//--- Construction
+
+impl<Octs> Key<Octs> {
+    /// Construct a new DNSSEC key manually.
+    pub fn new(owner: Name<Octs>, flags: u16, key: RawPublicKey) -> Self {
+        Self { owner, flags, key }
+    }
+}
+
+//--- Inspection
+
+impl<Octs> Key<Octs> {
+    /// The owner name attached to the key.
+    pub fn owner(&self) -> &Name<Octs> {
+        &self.owner
+    }
+
+    /// The flags attached to the key.
+    pub fn flags(&self) -> u16 {
+        self.flags
+    }
+
+    /// The raw public key.
+    pub fn raw_public_key(&self) -> &RawPublicKey {
+        &self.key
+    }
+
+    /// The signing algorithm used.
+    pub fn algorithm(&self) -> SecAlg {
+        self.key.algorithm()
+    }
+
+    /// Whether this is a zone signing key.
+    ///
+    /// From RFC 4034, section 2.1.1:
+    ///
+    /// > Bit 7 of the Flags field is the Zone Key flag.  If bit 7 has value
+    /// > 1, then the DNSKEY record holds a DNS zone key, and the DNSKEY RR's
+    /// > owner name MUST be the name of a zone.  If bit 7 has value 0, then
+    /// > the DNSKEY record holds some other type of DNS public key and MUST
+    /// > NOT be used to verify RRSIGs that cover RRsets.
+    pub fn is_zone_signing_key(&self) -> bool {
+        self.flags & (1 << 8) != 0
+    }
+
+    /// Whether this key has been revoked.
+    ///
+    /// From RFC 5011, section 3:
+    ///
+    /// > Bit 8 of the DNSKEY Flags field is designated as the 'REVOKE' flag.
+    /// > If this bit is set to '1', AND the resolver sees an RRSIG(DNSKEY)
+    /// > signed by the associated key, then the resolver MUST consider this
+    /// > key permanently invalid for all purposes except for validating the
+    /// > revocation.
+    pub fn is_revoked(&self) -> bool {
+        self.flags & (1 << 7) != 0
+    }
+
+    /// Whether this is a secure entry point.
+    ///
+    /// From RFC 4034, section 2.1.1:
+    ///
+    /// > Bit 15 of the Flags field is the Secure Entry Point flag, described
+    /// > in [RFC3757].  If bit 15 has value 1, then the DNSKEY record holds a
+    /// > key intended for use as a secure entry point.  This flag is only
+    /// > intended to be a hint to zone signing or debugging software as to
+    /// > the intended use of this DNSKEY record; validators MUST NOT alter
+    /// > their behavior during the signature validation process in any way
+    /// > based on the setting of this bit.  This also means that a DNSKEY RR
+    /// > with the SEP bit set would also need the Zone Key flag set in order
+    /// > to be able to generate signatures legally.  A DNSKEY RR with the SEP
+    /// > set and the Zone Key flag not set MUST NOT be used to verify RRSIGs
+    /// > that cover RRsets.
+    pub fn is_secure_entry_point(&self) -> bool {
+        self.flags & 1 != 0
```

Comment:
This didn't correctly detect the SEP flag for me when the input file contained a flags value of 257 (which should be ZONE key flag set AND SEP key flag set).

The following modified code worked for me:

```suggestion
        self.flags & 0b0000_0000_0000_0001 != 0
```